### PR TITLE
Current sensor component functionality

### PR DESCRIPTION
**Pull request by nitbharambe**
Relates to #547

### Changes proposed in this PR include:

- Component level functionality is added for current sensor

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/component/current_sensor.hpp [hunk 1]
```diff
@@ -117,29 +121,26 @@ template <symmetry_tag current_sensor_symmetry_> class CurrentSensor : public Ge
...
-    }
-    CurrentSensorCalcParam<asymmetric_t> asym_calc_param() const final {
-        CurrentSensorCalcParam<asymmetric_t> calc_param{};
-        // TODO
-        return calc_param;
+    template <symmetry_tag sym_calc> CurrentSensorCalcParam<sym_calc> calc_decomposed_param() const {
+        auto const i_polar = PolarComplexRDV<current_sensor_symmetry>(
+            {i_measured_, i_sigma_ * i_sigma_}, {i_angle_measured_, i_angle_sigma_ * i_angle_sigma_});
+        auto const i_decomposed = DecomposedComplexRDV<sym_calc>(i_polar);
+        return CurrentSensorCalcParam<sym_calc>{.angle_measurement_type = angle_measurement_type(),
```
Comment (nitbharambe) [inline power_grid_model_c/power_grid_model/include/power_grid_model/component/current_sensor.hpp:140]:
```suggestion
        auto const i_polar = PolarComplexRandVar<current_sensor_symmetry>(
            {i_measured_, i_sigma_ * i_sigma_}, {i_angle_measured_, i_angle_sigma_ * i_angle_sigma_});
        auto const i_decomposed = DecomposedComplexRandVar<sym_calc>(i_polar);
        return CurrentSensorCalcParam<sym_calc>{.angle_measurement_type = angle_measurement_type(),
```

**Review by figueroa1395 — COMMENTED** [review]
Looks good to me.

File: tests/cpp_unit_tests/test_current_sensor.cpp [hunk 2]
```diff
@@ -60,26 +66,28 @@ TEST_CASE("Test current sensor") {
...
-            CHECK(imag(asym_sensor_param.value[1]) == doctest::Approx(0.0));
-
-            CHECK(is_nan(sym_sensor_output_asym_param.id));
-            CHECK(is_nan(sym_sensor_output_asym_param.energized));
-            CHECK(is_nan(sym_sensor_output_asym_param.i_residual[0]));
-            CHECK(is_nan(sym_sensor_output_asym_param.i_angle_residual[1]));
+            CHECK(asym_sensor_param.i_real_variance[0] == doctest::Approx(i_variance_pu));
+            CHECK(asym_sensor_param.i_imag_variance[1] ==
+                  doctest::Approx(i_variance_pu * sin(deg_240) * sin(deg_240) +
+                                  i_angle_variance_pu * i_pu * i_pu * cos(deg_240) * cos(deg_240)));
```
[the remark this answers is not in the document]
Comment (nitbharambe) [reply tests/cpp_unit_tests/test_current_sensor.cpp:83]: Its better to expand it out here in the test than have this current sensor test dependent on statistics module. (Note: We depend on three_phase_tensor quite a lot in all our tests for readability. But here we have the choice to not do that.)

[the remark this answers is not in the document]
Comment (figueroa1395) [reply tests/cpp_unit_tests/test_current_sensor.cpp:83]: Is it though? I would argue the abstraction is better and I don't see why it matters if the test depends on the statistics module; it would actually be closer to how we would use this internally. However, I don't have a strong opinion, so feel free to resolve if you still think it is fine as is.

[the remark this answers is not in the document]
Comment (nitbharambe) [reply tests/cpp_unit_tests/test_current_sensor.cpp:83]: Tests generally need to be independently valid. If we sometime need to add/modify the statistics module, we would not want that to affect tests for the current sensor. We should definitely have all edge cases covered within statistics module itself. But this helps to isolate them. But on readability perspective, for a failing test, I should have knowledge only about current sensor component and not the entire statistics module. I understand we have most pitfalls covered for it to be okay to use the statistics module directly, but that's a good practice to avoid future issues.

[the remark this answers is not in the document]
Comment (figueroa1395) [reply tests/cpp_unit_tests/test_current_sensor.cpp:83]: You are totally right. This is a better practice indeed. Resolving.

File: tests/cpp_unit_tests/test_current_sensor.cpp [hunk 3]
```diff
@@ -30,7 +32,7 @@ TEST_CASE("Test current sensor") {
         for (auto const terminal_type :
              {MeasuredTerminalType::branch_from, MeasuredTerminalType::branch_to, MeasuredTerminalType::branch3_1,
               MeasuredTerminalType::branch3_2, MeasuredTerminalType::branch3_3}) {
-            CAPTURE(terminal_type);
+            // CAPTURE(terminal_type); // TODO check why this inerferes with normal output
```
[the remark this answers is not in the document]
Comment (figueroa1395) [reply tests/cpp_unit_tests/test_current_sensor.cpp]: Unresolved to check discrepancies in local machine. Todo with @nitbharambe

**Review by figueroa1395 — APPROVED** [review]

**Review by Jerry-Jinfeng-Guo — APPROVED** [review]
